raft/src/main/java/org/apache/kafka/raft/RaftClient.java

```diff
@@ -56,24 +57,17 @@
         void handleSnapshot(SnapshotReader<T> reader);
 
         /**
```

Comment:
It's fine to combine these APIs, but we need to document what happens if the current leader resigns, but we don't know who the new leader is yet. Do we get passed a LeaderAndEpoch with the current epoch + 1 and a node ID of -1? If so, do we then expect to see another LeaderAndEpoch with the current epoch + 1 and a valid node -1?

In other words, let's say node 0 is the leader and then resigns, and then node 1 becomes the leader. Does it look like this:
```
handleLeaderChange(LeaderAndEpoch(nodeId=0, epoch=0))
handleLeaderChange(LeaderAndEpoch(nodeId=-1, epoch=1))
handleLeaderChange(LeaderAndEpoch(nodeId=1, epoch=1))
```

Or would you rather have something like this?
```
handleLeaderChange(LeaderAndEpoch(nodeId=0, epoch=0))
handleLeaderChange(LeaderAndEpoch(nodeId=-1, epoch=0))
handleLeaderChange(LeaderAndEpoch(nodeId=1, epoch=1))
```

It seems like the second one will break a lot of invariants, so probably should be avoided. The first one might break some invariants too, though. We'd have to look.

Or you could choose to burn an epoch like this:
```
handleLeaderChange(LeaderAndEpoch(nodeId=0, epoch=0))
handleLeaderChange(LeaderAndEpoch(nodeId=-1, epoch=1))
handleLeaderChange(LeaderAndEpoch(nodeId=1, epoch=2))
```
Given that we only have a 31-bit epoch in the first place, that seems unwise, though.

Reply:
I'll update the documentation but in Raft epochs are not guarantee to have a leader. If there is a leader for an epoch then there is one and only one leader. So that means that the client could see.
```
handleLeaderChange(LeaderAndEpoch(nodeId=0, epoch=0))
handleLeaderChange(LeaderAndEpoch(nodeId=-1, epoch=1))
handleLeaderChange(LeaderAndEpoch(nodeId=1, epoch=1))
```
Or this for that matter
```
handleLeaderChange(LeaderAndEpoch(nodeId=0, epoch=0))
handleLeaderChange(LeaderAndEpoch(nodeId=-1, epoch=1))
handleLeaderChange(LeaderAndEpoch(nodeId=-1, epoch=2))
handleLeaderChange(LeaderAndEpoch(nodeId=1, epoch=2))
```

Reply:
> Or you could choose to burn an epoch like this...

This is not how leader election works in Raft. When a leader fails or steps down an epoch starts without a leader, only candidate(s). If leader election succeeds for a given epoch only one leader is guaranteed to elect for that epoch and will remain leader for the duration of that epoch.

Reply:
It sounds like the invariant here is that while `handleLeaderChange` may be called multiple times for the same epoch, only the last call will have a valid (non-negative) node id associated with it. Can you add a JavaDoc comment like this to the API?

Reply:
Yes. I added a few more sentences to make this clear.